lib/AST/GenericSignature.cpp

```diff
@@ -628,57 +628,51 @@ unsigned GenericSignatureImpl::getGenericParamOrdinal(
 Type GenericSignatureImpl::getNonDependentUpperBounds(Type type) const {
   assert(type->isTypeParameter());
 
+  auto &ctx = type->getASTContext();
+
   bool hasExplicitAnyObject = requiresClass(type);
 
   llvm::SmallVector<Type, 2> types;
+
+  // Class Inheritence
+  // If the class contains a type parameter that cannot be reduced,
+  // try looking for a non-dependent superclass.
   if (Type superclass = getSuperclassBound(type)) {
-    // If the class contains a type parameter, try looking for a non-dependent
-    // superclass.
-    while (superclass && superclass->hasTypeParameter()) {
-      superclass = superclass->getSuperclass();
+    if (auto *boundgeneric = dyn_cast<BoundGenericClassType>(superclass)) {
+
+      SmallVector<Type, 2> argTypes;
+
+      while (superclass && superclass->hasTypeParameter()) {
+
+        // check if the current class has generic arguments
+        for (Type argTy : boundgeneric->getGenericArgs()) {
+          argTypes.push_back(getReducedType(argTy));
```

Comment:
You can just call getReducedType(superclass) without taking it apart. getReducedType() recursively transforms the leaves (and if it didn't, it would be wrong to transform just the arguments anyway).